source/extensions/filters/http/jwt_authn/jwks_cache.cc

```diff
@@ -0,0 +1,180 @@
+#include "extensions/filters/http/jwt_authn/jwks_cache.h"
```

Comment:
I think you should push almost all of this code to jwt_verifier_lib, it doesn't have to belong to Envoy code base. The only dependency seems to be the configuration, but it is only audiences (vector<string>) and jwks (string) so you should be able to pass by parameter.

Reply:
ping

Reply:
I thought about it.  It is NOT just a pure cache.   It is a cached Jwks  + issuer_config.   For each request,  we need to access per-issuer config, such as payload_header,  and we also need to access cached Jwks.  We only need to do one map lookup by issuer if we combine jwks with issuer_config.  That is the current JwksCache design.   

If we split into a separate pure JwksCache,  then, I agree with you, it can be moved into Jwt_verify_lib. 
Current design is try to reduce a map lookup by issuer.

Reply:
issue config is just audience + jwks, no?

Reply:
also "forward_jwt" and "payload_header".  
On second throught,  I may go with your suggestion, it is cleaner.  I will create a config_map to lookup config by issuer, and jwks_cache.  If we use unorderred_map,  lookup is O(1),  it is OK to lookup twice.

Reply:
I tried, but decided to keep the current shape as:  with new JwtAuthn config change (https://github.com/envoyproxy/envoy/pull/3381), audience checking may not be in JwksCache object.  Further re-factory is needed to support the config change.